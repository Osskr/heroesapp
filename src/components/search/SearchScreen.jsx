import React, { useMemo } from 'react'
import { useLocation } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import { HeroCard } from '../heroes/HeroCard';
import queryString from 'query-string'
import { getHeroesByName } from '../../selectors/getHeroesByName';


export const SearchScreen = ({history}) => {
    
    //hook de react router dom para extraer la location
    const location = useLocation()
   
    const {q = '' } = queryString.parse(location.search)


    
    const [values, handleInputChange] = useForm({heroName: q})
    
    const {heroName} = values
    
    const handleSubmit = (e)=>{
        e.preventDefault();
        history.push(`?q=${heroName}`)
     }

     //const heroesFiltered = getHeroesByName(q)
    const heroesFiltered = useMemo(() => getHeroesByName(q), [q]) 
    return (
        <div>
            <h1>Search Screen</h1>

            <div className="row">
                <div className="col-4">
                    <h4>SearchForm</h4>
                    <hr />
                    <form onSubmit={handleSubmit}>
                        <input type="text"
                        name="heroName"
                        placeholder="find your hero"
                        className="form-control"
                        value = {q}
                        autoComplete="off"
                        onChange={handleInputChange}/>
                        <button type="submit" className="btn btn-outline-primary">Search</button>
                    </form>
                </div>
                <div className="col-8">
                <h4>Results</h4>
                <hr/>

                {
                    (q==='') 
                    &&
                    <div className="alert alert-info">
                        search a hero
                    </div>
                }
                {
                    heroesFiltered.map(hero => 
                        <HeroCard  key={hero.id}
                                    {...hero}/>
                    )
                }
            </div>

            </div>
        </div>
    )
}
