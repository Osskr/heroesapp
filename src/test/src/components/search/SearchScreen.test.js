import React from 'react'
import {mount} from 'enzyme'
import { SearchScreen } from '../../../../components/search/SearchScreen'
import { MemoryRouter, Route } from 'react-router-dom'



describe('Pruebas sobre el SearchScreen', () => {
  
  const wrapper = mount(<MemoryRouter initialEntries={['/search']}>
                            
                              <Route path='/search' component={SearchScreen}/>
                            
                        </MemoryRouter>)

  test('Se muestra con los valores por defecto', () => {

    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find('.alert-info').text().trim()).toBe('search a hero')
      
  })
  

  test('debe mostrar el heroe y el input con el valor del query string', () => {

    const wrapper = mount(<MemoryRouter initialEntries={['/search?q=batman']}>
                            
                              <Route path='/search' component={SearchScreen}/>
                            
                        </MemoryRouter>)

    expect(wrapper.find('input').prop('value')).toBe('batman')
    
      
  })

  test('debe llamar el push del history', () => {

    const history = {
      push:jest.fn()
    }

    const wrapper = mount(<MemoryRouter initialEntries={['/search?q=batman']}>
                            
                              <Route path='/search'
                                     component={()=> <SearchScreen history={history}/>}/>
                            
                        </MemoryRouter>)

    wrapper.find('input').simulate('change', {
      target:{
        name:'searchText',
        value:'batman'
      }
    })

    wrapper.find('form').prop('onSubmit')({
      preventDefault(){}
    })


    expect(history.push).toHaveBeenCalledWith('?q=batman')
      
  })

  
  
})
