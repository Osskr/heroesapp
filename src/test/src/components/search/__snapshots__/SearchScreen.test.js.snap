// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas sobre el SearchScreen Se muestra con los valores por defecto 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <div
    className="row"
  >
    <div
      className="col-4"
    >
      <h4>
        SearchForm
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="heroName"
          onChange={[Function]}
          placeholder="find your hero"
          type="text"
          value=""
        />
        <button
          className="btn btn-outline-primary"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-8"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        search a hero
      </div>
    </div>
  </div>
</div>
`;
